tests/MongoDB.Driver.Tests/PipelineDefinitionBuilderTests.cs

```diff
@@ -114,6 +114,61 @@ public void Merge_should_add_expected_stage()
             stages[0].Should().Be("{ $merge : { into : { db : 'database', coll : 'collection' } } }");
         }
 
+        [Fact]
+        public void UnionWith_should_add_expected_stage()
+        {
+            var pipeline = new EmptyPipelineDefinition<BsonDocument>();
+            var withCollection = Mock.Of<IMongoCollection<BsonDocument>>(
+                coll => coll.CollectionNamespace == CollectionNamespace.FromFullName("db.test"));
+            var withPipeline = new EmptyPipelineDefinition<BsonDocument>()
+                .AppendStage<BsonDocument, BsonDocument, BsonDocument>("{ $match : { b : 1 } }");
+
+            var result = pipeline.UnionWith<BsonDocument, BsonDocument, BsonDocument>(withCollection, withPipeline);
+
+            var stages = RenderStages(result, BsonDocumentSerializer.Instance);
+            stages[0].Should().Be("{ $unionWith : { coll : 'test', pipeline : [{ '$match' : { b : 1 } }] } }");
+        }
+
+        [Fact]
+        public void UnionWith_should_throw_when_pipeline_is_null()
+        {
+            PipelineDefinition<BsonDocument, BsonDocument> pipeline = null;
+            IMongoCollection<BsonDocument> withCollection = null;
+            var withPipeline = new EmptyPipelineDefinition<BsonDocument>();
+
+            var exception = Record.Exception(() => pipeline.UnionWith(withCollection, withPipeline));
+
+            var argumentNullException = exception.Should().BeOfType<ArgumentNullException>().Subject;
+            argumentNullException.ParamName.Should().Be("pipeline");
+        }
+
+        [Fact]
+        public void UnionWith_should_throw_when_TWith_is_not_the_same_with_TInput_and_pipeline_is_null()
+        {
+            var pipeline = new EmptyPipelineDefinition<BsonDocument>();
+            var withCollection = Mock.Of<IMongoCollection<BsonValue>>(
```

Comment:
I recommended:

    var withCollection = Mock.Of<IMongoCollection<object>>

because it is not really possible to have a collection of `BsonValue`  (MongoDB collections only work with documents, not BsonValues).